modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java

```diff
@@ -0,0 +1,221 @@
+        if (rawAfterKey != null) {
+            array.setTop(rawAfterKey.values);
+        }
+        this.keys = new TreeMap<>(array::compare);
+        this.canEarlyTerminate = Arrays.stream(sources)
+            .allMatch(CompositeValuesSourceConfig::canEarlyTerminate);
+    }
+
+    boolean canEarlyTerminate() {
+        return canEarlyTerminate;
+    }
+
+    private int[] getReverseMuls() {
+        return Arrays.stream(sources).mapToInt(CompositeValuesSourceConfig::reverseMul).toArray();
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long zeroBucket) throws IOException {
+        assert zeroBucket == 0L;
+
+        // Replay all documents that contain at least one top bucket (collected during the first pass).
+        grow(keys.size()+1);
+        for (LeafContext context : contexts) {
+            DocIdSetIterator docIdSetIterator = context.docIdSet.iterator();
+            if (docIdSetIterator == null) {
+                continue;
+            }
+            final LeafBucketCollector collector = array.getLeafCollector(context.ctx, getSecondPassCollector(context.subCollector));
+            int docID;
+            while ((docID = docIdSetIterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                collector.collect(docID, 0L);
+            }
+        }
+
+        int num = Math.min(size, keys.size());
+        final InternalComposite.InternalBucket[] buckets = new InternalComposite.InternalBucket[num];
+        final int[] reverseMuls = getReverseMuls();
+        int pos = 0;
+        for (int slot : keys.keySet()) {
+            CompositeKey key = array.toCompositeKey(slot);
+            InternalAggregations aggs = bucketAggregations(slot);
+            int docCount = bucketDocCount(slot);
+            buckets[pos++] = new InternalComposite.InternalBucket(key, reverseMuls, docCount, aggs);
+        }
+        return new InternalComposite(name, size, Arrays.asList(buckets), reverseMuls, pipelineAggregators(), metaData());
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        final int[] reverseMuls = getReverseMuls();
+        return new InternalComposite(name, size, Collections.emptyList(), reverseMuls, pipelineAggregators(), metaData());
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (leaf != null) {
+            leaf.docIdSet = builder.build();
+            contexts.add(leaf);
+        }
+        leaf = new LeafContext(ctx, sub);
+        builder = new RoaringDocIdSet.Builder(ctx.reader().maxDoc());
+        return array.getLeafCollector(ctx, getFirstPassCollector());
+    }
+
+    @Override
+    protected void doPostCollection() throws IOException {
+        if (leaf != null) {
+            leaf.docIdSet = builder.build();
+            contexts.add(leaf);
+        }
+    }
+
+    private LeafBucketCollector getFirstPassCollector() {
```

Comment:
I think this method and the next method can just return `LeafCollector`? 

The bucket ord isn't used in this implementations. Also these collectors are not directly used by the aggs framework, but are wrapped by a `LeafBucketCollector` instance in `CompositeValuesSource.java`.